src/gululog_idx.erl

```diff
@@ -285,39 +315,172 @@ truncate(Dir, #idx{tid = Tid, fd = Fd} = Idx, SegId, LogId, BackupDir) ->
 
 %%%*_ PRIVATE FUNCTIONS ========================================================
 
+-spec to_location(false | entry()               ) -> false | location().
+to_location(                               false) -> false;
+to_location(?ENTRY(_LogId, _Ts, SegId, Position)) -> {SegId, Position}.
+
+%% @private Read entry from cache, read from file if deleted from cache.
+-spec read_entry(dirname(), index() | cache(), logid()) -> false | entry().
+read_entry(Dir, #idx{tid = Tid}, LogId) -> read_entry(Dir, Tid, LogId);
+read_entry(Dir, Tid, LogId) ->
+  not is_out_of_range(Tid, LogId) andalso
+  case lookup_cache(Tid, LogId) of
+    false -> read_file_entry(Dir, get_segid(Tid, LogId), LogId);
+    Entry -> Entry
+  end.
+
 %% @private Truncate cache, from the given logid (inclusive).
--spec truncate_cache(cache(), logid()) -> cache().
-truncate_cache(Tid, LogId) ->
-  Ms = ets:fun2ms(fun(?ETS_ENTRY(_, LogIdX, _)) -> LogIdX >= LogId end),
-  _ = ets:select_delete(Tid, Ms),
+%% re-initialize the cache for last segment in case there are entries deleted.
+%% @end
+-spec truncate_cache(cache(), filename(), logid()) -> cache().
+truncate_cache(Tid, NewSegIdxFile, LogId) ->
+  Ms = ets:fun2ms(fun(?ENTRY(LogIdX, _, _, _)) -> LogIdX >= LogId end),
+  _  = ets:select_delete(Tid, Ms),
+  %%   To make sure that the latest entry is always in cache.
+  ok = init_cache_from_file(NewSegIdxFile, Tid),
```

Comment:
Yes. the 'last/latest' entry in cache might have been deleted before truncation, it's important to recover the real 'last/latest'.
we had no test case to ensure it, perhaps add one ?